### PR TITLE
kvserver: document reproposals

### PR DESCRIPTION
**Pull request by tbg**
Reproposals are a deep rabbit hole and an area in which past changes
were all related to subtle bugs. Write it all up and in particular make
some simplifications that ought to be possible if my understanding is
correct:

- have proposals always enter `(*Replica).propose` without a
  MaxLeaseIndex or prior encoded command set, i.e. `propose`
  behaves the same for reproposals as for first proposals.
- assert that after a failed call to tryReproposeWithNewLeaseIndex,
  the command is not in the proposals map, i.e. check absence of
  a leak.
- replace code that should be impossible to reach (and had me confused
  for a good amount of time) with an assertion.
- add long comment on `r.mu.proposals`.

This commit also moves `tryReproposeWithNewLeaseIndex` off `(*Replica)`,
which is possible due to recent changes[^1]. In doing so, I realized
there was a (small) data race (now fixed): when returning a
`NotLeaseholderError` from that method, we weren't acquiring `r.mu`. It
may have looked as though we were holding it already since we're
accessing `r.mu.propBuf`, however that field has special semantics - it
wraps `r.mu` and acquires it when needed.

[^1]: The "below raft" test mentioned in the previous comment was
changed in https://github.com/cockroachdb/cockroach/pull/93785 and
no longer causes a false positive.

Epic: CRDB-220
Release note: None

### CODE REVIEW

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @pavelkalinnikov, and @tbg)

___
*[`pkg/kv/kvserver/replica.go` line 506 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLR_ILV7lAt5WD7dHtd:-NLR_ILV7lAt5WD7dHte:bd5d6lg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica.go#L506)):*
> ```Go
> 		//    is not superseded, see below) and attach the value to the entry.
> 		// 2. for each entry:
> 		//				- stage written and in-memory effects of the entry (some may apply as no-ops
> ```

indentation
___
*[`pkg/kv/kvserver/replica.go` line 596 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLR_GhY2i3ZaI7ZdBBi:-NLR_GhZ3LfJF0p7p0rF:b-1ko7fm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica.go#L596)):*
> ```Go
> 		// [^1]: https://github.com/cockroachdb/cockroach/blob/59ce13b6052a99a0318e3dfe017908ff5630db30/pkg/kv/kvserver/replica_raft.go#L1224
> 		// [^2]: https://github.com/cockroachdb/cockroach/blob/59ce13b6052a99a0318e3dfe017908ff5630db30/pkg/kv/kvserver/replica_application_result.go#L148
> 		// [^3]: it's debatable how useful this is. It was introduced in
> ```

what is "this"?

**Review by nvb — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @tbg)

___
*[`pkg/kv/kvserver/replica.go` line 523 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRofkDE-tISCnfPHqa:-NLRofkDE-tISCnfPHqb:b-ld3f0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica.go#L523)):*
> ```Go
> 		// We can always safely create an identical copy (i.e. (1)) because of the
> 		// replay protection conferred by the MaxLeaseIndex - all but the first
> 		// proposal will be rejected (i.e. apply as a no-op).
> ```

"all but the first proposal that successfully achieves consensus"
___
*[`pkg/kv/kvserver/replica.go` line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRp8QK-K6ya4TclPOs:-NLRp8QK-K6ya4TclPOt:b-8hycas) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica.go#L531)):*
> with various different MaxLeaseIndex values

Is it possible for there two be more than two different MaxLeaseIndex values in the unapplied log at a time? And can they be interleaved?
___
*[`pkg/kv/kvserver/replica.go` line 552 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRpWzx6StobUfgq0iC:-NLRpWzx6StobUfgq0iD:bktv5so) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica.go#L552)):*
> ```Go
> 		// where we assume that the `MaxLeaseIndex` 100 is invalid, i.e. when we see the
> 		// first copy of the command being applied, we've already applied some command
> 		// with a higher `MaxLeaseIndex`. In a world without mechanism (2), `N` would
> ```

"an equal or higher"
___
*[`pkg/kv/kvserver/replica.go` line 553 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRpZtmESJ5D4AphWQD:-NLRpZtmESJ5D4AphWQE:bifcn3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica.go#L553)):*
> ```Go
> 		// first copy of the command being applied, we've already applied some command
> 		// with a higher `MaxLeaseIndex`. In a world without mechanism (2), `N` would
> 		// be rejected, and would finalize the proposal (i.e. signal the client with
> ```

Should we define "finalize" up above in a broader discussion of when entries are added to the proposals map, when they're updated, and when they're removed?
___
*[`pkg/kv/kvserver/replica.go` line 584 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRq8Dx3Arr7cNwdxTF:-NLRq8Dx3Arr7cNwdxTG:b-yeu3y2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica.go#L584)):*
> ```Go
> 		// then can (2) engage. In addition, an entry that doesn't pass this equality
> 		// check must not signal the proposer and/or unlink from the proposals map (as a
> 		// newer reproposal is likely in the log)[^4].
> ```

"as a newer reproposal is likely in the log which may succeed"
___
*[`pkg/kv/kvserver/replica_application_result.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRrYWt2iHKO5WoBrpO:-NLRrYWt2iHKO5WoBrpP:b-gkazo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica_application_result.go#L259)):*
> ```Go
> 	p.command.MaxLeaseIndex = 0
> 	p.encodedCommand = nil
> 	pErr := r.propose(ctx, p, tok.Move(ctx))
> ```

`Replica.propose` doesn't assign the MaxLeaseIndex synchronously, it only does when the proposal buffer is flushed. Are we ok with the command having a MaxLeaseIndex of 0 in the meantime? The need to reset the fields on error raises questions.
___
*[`pkg/kv/kvserver/replica_application_state_machine.go` line 236 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRtNrQCiQWNX3y1GQg:-NLRtNrQCiQWNX3y1GQh:b-il6d5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1233022fc2f67a3901c0be08ec71abb13b1e8fe/pkg/kv/kvserver/replica_application_state_machine.go#L236)):*
> ```Go
> 			log.Fatalf(ctx, "command already applied: %+v; unexpected successful result", cmd)
> 		}
> 		// If any reproposals at a higher MaxLeaseIndex exist we know that they will
> ```

I don't understand this branch. It dates back to https://github.com/cockroachdb/cockroach/commit/0e81674e1f7af3578155d44451a756c5a88e85b1, but I can't make sense of it — it sounds like it was just dead code. Glad you got rid of it and replaced it with an assertion.
___
*[`pkg/kv/kvserver/replica_proposal.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRtpMY9lN48Tp4WOIM:-NLRtpMY9lN48Tp4WOIN:b-ht2vvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica_proposal.go#L135)):*
> ```Go
> // rejected and should not be reproposed.
> //
> // Note that some commands (such as leases) don't use MaxLeaseIndex.
> ```

"lease requests"

Lease transfers do use MaxLeaseIndex.

**Review by tbg — COMMENTED** [review]
Thanks for the reviews! I rebased and added a fix-up commit, so if reviewable gets confused you might just want to use GH review for the last commit instead.

> Dismissed @andreimatei from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-:-NLel-aZ4s2PplbyUyDh:b9r2zge)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/replica.go` line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRp8QK-K6ya4TclPOs:-NLeg6vZ1TWQMLCYcCtS:b-yp8xm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica.go#L531)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> with various different MaxLeaseIndex values

Is it possible for there two be more than two different MaxLeaseIndex values in the unapplied log at a time? And can they be interleaved?
</blockquote></details>

I think in practice we won't see too much of that but in theory anything goes, because you always have to assume that all previous incarnations are still floating around out there as raft reproposals that could reach the leader at any time & in any order. Updated the comment with some detail and simplified it a bit too.
___
*[`pkg/kv/kvserver/replica.go` line 596 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLR_GhY2i3ZaI7ZdBBi:-NLekGX4BeW5akc8_abL:buvnjiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica.go#L596)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what is "this"?
</blockquote></details>

The below-raft reproposal mechanism, updated.
___
*[`pkg/kv/kvserver/replica_application_result.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRrYWt2iHKO5WoBrpO:-NLekNek4T9xOSy_uX06:b-szbkkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica_application_result.go#L259)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`Replica.propose` doesn't assign the MaxLeaseIndex synchronously, it only does when the proposal buffer is flushed. Are we ok with the command having a MaxLeaseIndex of 0 in the meantime? The need to reset the fields on error raises questions.
</blockquote></details>

The command is not in the map at this point, so I am not only ok with this, but intentionally want this. I want `r.propose` to not have to care about whether it sees a proposal for the first time or fifth time.
___
*[`pkg/kv/kvserver/replica_application_state_machine.go` line 236 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRtNrQCiQWNX3y1GQg:-NLekgBD9jC9VvJmMRsx:b-pbaqg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1233022fc2f67a3901c0be08ec71abb13b1e8fe/pkg/kv/kvserver/replica_application_state_machine.go#L236)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't understand this branch. It dates back to https://github.com/cockroachdb/cockroach/commit/0e81674e1f7af3578155d44451a756c5a88e85b1, but I can't make sense of it — it sounds like it was just dead code. Glad you got rid of it and replaced it with an assertion.
</blockquote></details>

Ack, thanks for the check. I had a similar conclusion. The logic is still quite complicated and I wish it were a little less subtle to know when a proposal is in which "state" but that's for another day.

**Review by nvb — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/kv/kvserver/replica_application_result.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRrYWt2iHKO5WoBrpO:-NMZhkYeC7a4SaFTf2ic:b-rqhusk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica_application_result.go#L259)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The command is not in the map at this point, so I am not only ok with this, but intentionally want this. I want `r.propose` to not have to care about whether it sees a proposal for the first time or fifth time.
</blockquote></details>

The part that feels fragile to me then is the need to reset the fields on error. I probably just don't understand that well enough yet, but why is an error here treated differently than an error when flushing the proposal buffer back into the `proposals` map?

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94633)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/kv/kvserver/replica_application_result.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRrYWt2iHKO5WoBrpO:-NNN9dq4BhDyWoRitBUo:bjh9fb7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica_application_result.go#L259)):*
In this method, the command is not in the proposals map. If `r.propose` returns an error, the command never made it into the proposals map. So here we just pretend that we never even called `r.propose`. The proposal will then fail up to the caller, and this is probably not a particularly well-tested code path anyway, but it will not be reproposed and so it shouldn't matter what the MLAI and prev encoded command was. But it felt right to restore the previous state, to avoid having to reason about yet one more code path at the caller.

I'm not defending this code, btw - this is the clearest I was able to make it without really stirring the pot. I wish we didn't have these below-raft reproposals and would "just" delegate to an above-raft goroutine pool to repropose on behalf of clients who used pipelined writes. I bet that would be tricky in its own right, but conceptually clearer.

> but why is an error here treated differently than an error when flushing the proposal buffer back into the proposals map?

I might be wrong, but: if we catch an error flushing the buffer, we return it up:

https://github.com/cockroachdb/cockroach/blob/a94858bff9a53450e0c76ff8ed8757fd3d18a264/pkg/kv/kvserver/replica_proposal_buf.go#L540

and up:

https://github.com/cockroachdb/cockroach/blob/b637f0301300c769c4f7f51724a38bbab1d03cbb/pkg/kv/kvserver/replica_raft.go#L727-L728

and up:

https://github.com/cockroachdb/cockroach/blob/b637f0301300c769c4f7f51724a38bbab1d03cbb/pkg/kv/kvserver/replica_raft.go#L756-L758

and up:

https://github.com/cockroachdb/cockroach/blob/b637f0301300c769c4f7f51724a38bbab1d03cbb/pkg/kv/kvserver/replica_raft.go#L1110-L1113

basically flushing the buffer "has to work". But in this method, we can run into the closed timestamp, so it has to be fallible and we need to return an error to the caller.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-:-NNlkLimE-BAHO3Z_xrI:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/kv/kvserver/replica_application_result.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRrYWt2iHKO5WoBrpO:-NNlkMuICgnM8tchL4l8:ban8jjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica_application_result.go#L259)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

In this method, the command is not in the proposals map. If `r.propose` returns an error, the command never made it into the proposals map. So here we just pretend that we never even called `r.propose`. The proposal will then fail up to the caller, and this is probably not a particularly well-tested code path anyway, but it will not be reproposed and so it shouldn't matter what the MLAI and prev encoded command was. But it felt right to restore the previous state, to avoid having to reason about yet one more code path at the caller.

I'm not defending this code, btw - this is the clearest I was able to make it without really stirring the pot. I wish we didn't have these below-raft reproposals and would "just" delegate to an above-raft goroutine pool to repropose on behalf of clients who used pipelined writes. I bet that would be tricky in its own right, but conceptually clearer.

> but why is an error here treated differently than an error when flushing the proposal buffer back into the proposals map?

I might be wrong, but: if we catch an error flushing the buffer, we return it up:

https://github.com/cockroachdb/cockroach/blob/a94858bff9a53450e0c76ff8ed8757fd3d18a264/pkg/kv/kvserver/replica_proposal_buf.go#L540

and up:

https://github.com/cockroachdb/cockroach/blob/b637f0301300c769c4f7f51724a38bbab1d03cbb/pkg/kv/kvserver/replica_raft.go#L727-L728

and up:

https://github.com/cockroachdb/cockroach/blob/b637f0301300c769c4f7f51724a38bbab1d03cbb/pkg/kv/kvserver/replica_raft.go#L756-L758

and up:

https://github.com/cockroachdb/cockroach/blob/b637f0301300c769c4f7f51724a38bbab1d03cbb/pkg/kv/kvserver/replica_raft.go#L1110-L1113

basically flushing the buffer "has to work". But in this method, we can run into the closed timestamp, so it has to be fallible and we need to return an error to the caller.
</blockquote></details>

Thanks for explaining. This makes sense. If you wanted to add another sentence to the comment stating that an error to propose causes the request to fail while a success means that it's back in the proposal pipeline, I'd be supportive. It's also fine if not.

**Review by tbg — COMMENTED** [review]
bors r=nvanbenschoten

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-:-NNlyw_mDKLcSCWO-bKZ:b-nvg7rs)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/replica_application_result.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94633#-NLRrYWt2iHKO5WoBrpO:-NNlyvIJ8cZPGbEQqG33:b-a60o3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4acc711a955b33a23d0de8c099d9b1bd3a3df5/pkg/kv/kvserver/replica_application_result.go#L259)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Thanks for explaining. This makes sense. If you wanted to add another sentence to the comment stating that an error to propose causes the request to fail while a success means that it's back in the proposal pipeline, I'd be supportive. It's also fine if not.
</blockquote></details>

Good idea., done.
